### PR TITLE
Fix up support for changing rule set files through the Solution Explorer 

### PR DESCRIPTION
**Pull request by jasonmalinowski**
The AnalyzersCommandHandler was still trying to call GetHostProject to find the host project as a way to directly grab the rule set file to process that. That was deprecated but we hadn't fixed it up yet.

The fix is to re-thread through a way to get to the rule set. The "ownership" of a rule set file (if one exists for a project) is owned by the VisualStudioProjectOptionsProcessor, so we now need a way to jump to that directly from any project so another map is added to the VisualStudioWorkspaceImpl. That map is unfortunate -- the OptionsProcessor is really something that should be private to the project system itself and shouldn't be grabbable by any random code.

Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/779916
Completes task https://devdiv.visualstudio.com/DevDiv/_workitems/edit/698029

<details><summary>Ask Mode template</summary>

### Customer scenario

If you go to the Solution Explorer and expand the Analyzers node, you can browse the rules that are associated with the analyzers you reference. If you right click a rule, choose Set Severity, and choose an option, you get an error message instead of it working.

### Bugs this fixes

Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/779916

### Workarounds, if any

Just edit the rule set file with the designer or manually.

### Risk

Low.

### Performance impact

None, nothing is really changing.

### Is this a regression from a previous update?

Yes, this was broken in Dev16.0 Preview 1.

### Root cause analysis

This was an incomplete refactoring that we had a TODO comment for but the tracking bug got misassigned. It was then caught in dogfooding and fixed. We've also figured out where the bug went and are cleaning those up.

### How was the bug found?

Internal dogfooding.

</details>

### CODE REVIEW

**Comment by jinujoseph** [comment]
@jasonmalinowski pls retarget this to master.

**Comment by jasonmalinowski** [comment]
Tagging @dotnet/roslyn-ide for review.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs [hunk 1]
```diff
@@ -63,6 +63,14 @@ internal abstract partial class VisualStudioWorkspaceImpl : VisualStudioWorkspac
 
         private ImmutableDictionary<ProjectId, IVsHierarchy> _projectToHierarchyMap = ImmutableDictionary<ProjectId, IVsHierarchy>.Empty;
         private ImmutableDictionary<ProjectId, Guid> _projectToGuidMap = ImmutableDictionary<ProjectId, Guid>.Empty;
+
+        /// <summary>
+        /// A map to fetch the <see cref="VisualStudioProjectOptionsProcessor"/> for a project. This right now is only used to implement
+        /// <see cref="TryGetRuleSetForProject(ProjectId)"/> and any other use is extremely suspicious, since direct use of this is out of
+        /// sync with the Workspace if there is active batching happening.
+        /// </summary>
+        private Dictionary<ProjectId, VisualStudioProjectOptionsProcessor> _projectToOptionsProcessorMap = new Dictionary<ProjectId, VisualStudioProjectOptionsProcessor>();
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: :bulb: If it's not removed, it seems reasonable to restrict the exposure of the options processor in two ways:

1. First switch to `Func<IRuleSetFile>` as the value, which only allows the workspace access to the rule set file.
2. Switch from `Func<IRuleSetFile>` &rarr; `Func<string>` (rule set file path), which further restricts what the workspace can do with that object _and_ resolves the lifetime concerns described in https://github.com/dotnet/roslyn/pull/32990#discussion_r253169866.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs [hunk 2]
```diff
@@ -213,6 +229,17 @@ internal AbstractProject GetHostProject(ProjectId projectId)
             return new StubProject(ProjectTracker, project, GetHierarchy(projectId), project.OutputFilePath);
         }
 
+        // TODO: consider whether this should be going to the project system directly to get this path. This is only called on interactions from the
```
[the remark this answers is not in the document]
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: I _can_, but it'll be another bug that'll never get looked at in our pile of bugs that's already too big to manage. So...?

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs [hunk 3]
```diff
@@ -136,6 +144,14 @@ internal void AddProjectToInternalMaps(VisualStudioProject project, IVsHierarchy
             }
         }
 
+        internal void AddProjectOptionsProcessorToInternalMaps(VisualStudioProject project, VisualStudioProjectOptionsProcessor optionsProcessor)
+        {
+            lock (_gate)
+            {
+                _projectToOptionsProcessorMap.Add(project.Id, optionsProcessor);
```
[the remark this answers is not in the document]
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: No: that'd mean we'd have two projects with the same Id and that'd be most bad.

**Comment by jasonmalinowski** [comment]
@jinujoseph This is ready for M2 approval.

**Pull request merged**